### PR TITLE
Update type annotations

### PR DESCRIPTION
**Pull request by matthewdouglas**
Some improvements to type annotations as best I could work out.

### CODE REVIEW

File: bitsandbytes/autograd/_functions.py [hunk 1]
```diff
@@ -106,7 +106,14 @@ def undo_layout(permuted_tensor: torch.Tensor, tile_indices: torch.LongTensor) -
...
 class MatMul8bit(torch.autograd.Function):
     @staticmethod
-    def forward(ctx, A, B, out=None, quant_type="vector", precision=None):
+    def forward(
+        ctx,
+        A: torch.Tensor,
+        B: torch.Tensor,
+        out=None,
+        quant_type: F.VectorwiseQuantType = "vector",
+        precision=Optional[List[int]],
```
Comment (akx) [inline bitsandbytes/autograd/_functions.py:115]: ⚠️ Type annotation had become the default value here?
```suggestion
        precision: Optional[List[int]] = None,
```

File: bitsandbytes/autograd/_functions.py [hunk 2]
```diff
@@ -490,7 +497,9 @@ class MatMul4Bit(torch.autograd.Function):
     # backward is mostly the same, but adds one extra clause (see "elif state.CxB is not None")
 
     @staticmethod
-    def forward(ctx, A, B, out=None, bias=None, quant_state: Optional[F.QuantState] = None):
+    def forward(
+        ctx, A: torch.Tensor, B: torch.Tensor, out=None, bias=None, quant_state: Optional[F.QuantState] = None
```
Comment (akx) [inline bitsandbytes/autograd/_functions.py:501]: Add the trailing comma so Ruff will spread this out onto multiple lines for ease of reading? (Will repeat this as "trailing comma?" throughout this review)
```suggestion
        ctx, A: torch.Tensor, B: torch.Tensor, out=None, bias=None, quant_state: Optional[F.QuantState] = None,
```

File: bitsandbytes/functional.py [hunk 3]
```diff
@@ -202,8 +202,10 @@ def prefetch_tensor(A, to_cpu=False):
     lib.cprefetch(get_ptr(A), ct.c_size_t(num_bytes), ct.c_int32(deviceid))
 
 
-def elementwise_func(func_name, A, B, value, prefetch=True):
-    func = None
+def elementwise_func(
+    func_name: str, A: torch.Tensor, B: Optional[torch.Tensor], value: Union[float, int], prefetch=True
```
Comment (akx) [inline bitsandbytes/functional.py:206]: Trailing comma?

File: bitsandbytes/functional.py [hunk 4]
```diff
@@ -454,17 +456,22 @@ def get_ptr(A: Optional[Tensor]) -> Optional[ct.c_void_p]:
...
+def post_call(prev_device: Union[int, str, torch.device]):
     torch.cuda.set_device(prev_device)
 
 
-def get_transform_func(dtype, orderA, orderOut, transpose=False):
+TransformOrder = Literal["row", "col", "col32", "col_turing", "col_ampere"]
+
+
+def get_transform_func(
+    dtype: torch.dtype, orderA: TransformOrder, orderOut: TransformOrder, transpose=False
```
Comment (akx) [inline bitsandbytes/functional.py:473]: Trailing comma?

File: bitsandbytes/functional.py [hunk 5]
```diff
@@ -1013,7 +1027,10 @@ def dequantize_blockwise(
     return out
 
 
-def get_4bit_type(typename, device=None, blocksize=64):
+DataType4Bit = Literal["nf4", "fp4", "int4", "af4"]
+
+
+def get_4bit_type(typename: DataType4Bit, device: Optional[Union[int, str, torch.device]] = None, blocksize=64):
```
Comment (akx) [inline bitsandbytes/functional.py:1033]: Trailing comma, type for `blocksize`?

File: bitsandbytes/functional.py [hunk 6]
```diff
@@ -2592,7 +2631,15 @@ def double_quant(A, col_stats=None, row_stats=None, out_col=None, out_row=None,
...
 
-def transform(A, to_order, from_order="row", out=None, transpose=False, state=None, ld=None):
+def transform(
+    A: torch.Tensor,
+    to_order: TransformOrder,
+    from_order: TransformOrder = "row",
+    out: Optional[torch.Tensor] = None,
+    transpose=False,
+    state: Optional[Tuple[torch.Size, TransformOrder]] = None,
+    ld=None,
```
Comment (akx) [inline bitsandbytes/functional.py:2641]: Type?

File: bitsandbytes/functional.py [hunk 7]
```diff
@@ -2772,8 +2824,14 @@ def spmm_coo_very_sparse(cooA, B, dequant_stats=None, out=None):
 
 C = 127.0
 
+VectorwiseQuantType = Literal[
+    "linear", "vector", "row", "zeropoint", "vector-zeropoint", "row-zeropoint", "truncated-vector"
```
Comment (akx) [inline bitsandbytes/functional.py:2828]: Trailing comma?

File: bitsandbytes/functional.py [hunk 8]
```diff
@@ -2772,8 +2824,14 @@ def spmm_coo_very_sparse(cooA, B, dequant_stats=None, out=None):
...
+]
+
 
-def vectorwise_quant(x, dim=1, quant_type="vector"):
+def vectorwise_quant(
+    x: torch.Tensor, dim=1, quant_type: VectorwiseQuantType = "vector"
```
Comment (akx) [inline bitsandbytes/functional.py:2833]: Trailing comma?

File: bitsandbytes/functional.py [hunk 9]
```diff
@@ -2817,15 +2875,17 @@ def vectorwise_quant(x, dim=1, quant_type="vector"):
...
     if quant_type == "vector":
         x = (xq / C * max1).to(torch.float32)
         return x
     else:
         return None
 
 
-def vectorwise_mm_dequant(xq, S1, S2, dtype=torch.half, quant_type="vector"):
+def vectorwise_mm_dequant(
+    xq: torch.Tensor, S1: torch.Tensor, S2: torch.Tensor, dtype=torch.half, quant_type: VectorwiseQuantType = "vector"
```
Comment (akx) [inline bitsandbytes/functional.py:2887]: Trailing comma?

File: bitsandbytes/functional.py [hunk 10]
```diff
@@ -2884,7 +2944,9 @@ def vectorwise_mm_dequant(xq, S1, S2, dtype=torch.half, quant_type="vector"):
         return None
 
 
-def dequant_min_max(xq, A, B, SA, SB, dtype=torch.half):
+def dequant_min_max(
+    xq: torch.Tensor, A: torch.Tensor, B: torch.Tensor, SA: torch.Tensor, SB: torch.Tensor, dtype=torch.half
```
Comment (akx) [inline bitsandbytes/functional.py:2948]: Trailing comma?

**Review by akx — CHANGES_REQUESTED** [review]

File: bitsandbytes/autograd/_functions.py (hunk 1 above)
Comment (matthewdouglas) [reply bitsandbytes/autograd/_functions.py:115]: Good catch, thanks!

File: bitsandbytes/functional.py (hunk 5 above)
Comment (matthewdouglas) [reply bitsandbytes/functional.py:1033]: `blocksize` should be inferred as `int` here.

File: bitsandbytes/functional.py [hunk 11]
```diff
@@ -2252,7 +2269,7 @@ def batched_igemm(
     return out
 
 
-def igemmlt(A, B, SA, SB, out=None, Sout=None, dtype=torch.int32):
+def igemmlt(A: torch.Tensor, B: torch.Tensor, SA, SB, out=None, Sout=None, dtype=torch.int32):
```
Comment (matthewdouglas) [inline bitsandbytes/functional.py:2272]: TODO: Need to make a type alias for `SA` and `SB`

File: bitsandbytes/functional.py (hunk 6 above)
Comment (matthewdouglas) [reply bitsandbytes/functional.py:2641]: I'm not sure if `ld` belongs here at all considering it's not used, but it is in `nvidia_transform`

**Comment by Titus-von-Koeller** [comment]
This is still in draft state, right? @matthewdouglas

No rush, but happy to merge once ready.

**Comment by matthewdouglas** [comment]
Superseded by newer PRs.

**Pull request closed without merging**